dbms/src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3807,4 +3830,88 @@ bool MergeTreeData::moveParts(CurrentlyMovingPartsTagger && moving_tagger)
+    Poco::File file(inserted_parts_path);
+    bool created = file.createFile();
+    if (created)
+    {
+        LOG_INFO(log, "Created " << inserted_parts_path);
+    }
+
+    // Fix corrupted file when necessary
+    size_t file_size = file.getSize();
+    if (file_size % sizeof(UInt128) != 0)
+    {
+        size_t exp_file_size = (file_size % sizeof(UInt128))*sizeof(UInt128);
+        file.setSize(exp_file_size);
+        LOG_ERROR(log, inserted_parts_path << " size " << file_size << " is incorrect. Truncated it to " << exp_file_size << ".");
+        file_size = exp_file_size;
+    }
+
+    // Shorten file when necessary
+    size_t file_size_limit = std::max(size_t(1000000000UL), size_t(2 * getSettings()->insertion_deduplication_parts * sizeof(UInt128)));
+    if (file_size > file_size_limit)
+    {
+        size_t new_size = getSettings()->insertion_deduplication_parts * sizeof(UInt128) / 2;
+        int r = shorten(inserted_parts_path.data(), new_size);
+        if (r != 0)
+        {
+            LOG_WARNING(log, "Cannot shorten " << inserted_parts_path
+                << " from " << file_size << " to " << new_size
+                <<". error: " << strerror(errno));
+            file_size = file.getSize();
+        }
+        else
+        {
+            LOG_INFO(log, "Has shorten " << inserted_parts_path << " from " << file_size << " to " << new_size);
+            file_size = new_size;
+        }
+    }
+
+    off_t off = 0;
+    if (file_size > getSettings()->insertion_deduplication_parts * sizeof(UInt128) / 2)
+        off = file_size - getSettings()->insertion_deduplication_parts * sizeof(UInt128) / 2;
+    ReadBufferFromFile inserted_parts_in_file(inserted_parts_path);
+    inserted_parts_in_file.seek(off, SEEK_SET);
+
+    size_t size = size_t(getSettings()->insertion_deduplication_parts * getSettings()->insertion_deduplication_bits_per_part / 8);
+    size_t hashs = std::round(double(getSettings()->insertion_deduplication_bits_per_part) * std::log(2.0));
+    inserted_parts_filter = std::make_unique<BloomFilter>(size, hashs, rand());
+
+    size_t n = (file_size - off) / sizeof(UInt128);
+    UInt128 checksum;
+    for (size_t i=0; i<n; i++)
+    {
+        readBinary(checksum, inserted_parts_in_file);
+        inserted_parts_filter->add(reinterpret_cast<const char *>(&checksum), sizeof(UInt128));
+    }
+    // Write-through to reduce corruption risk
+    inserted_parts_file = std::make_unique<WriteBufferFromFile>(inserted_parts_path, 4096, O_WRONLY | O_APPEND | O_CREAT | O_SYNC);
+    inserted_parts_num = n;
+    LOG_DEBUG(log, "Loaded parts insertion history (" << n << " items)");
+}
+
+bool MergeTreeData::findPartInInsertionHistory(UInt128 part_data_checksum)
+{
+    std::lock_guard lock(inserted_parts_mutex);
+    if (inserted_parts_filter==nullptr)
+        loadPartsInsertionHistory();
+    bool found = inserted_parts_filter->find(reinterpret_cast<const char *>(&part_data_checksum), sizeof(UInt128));
```

Comment:
Approximate filtering - is it a good use case for all users?
I think, we can simply store latest N hashes in hash table in memory.
(In two hash tables. And drop first hash table and replace with second when it becomes full.)

The memory usage will be moderate even for 100 000 000 data parts. That is usually way more than needed. (In contrast, for ReplicatedMergeTree we store only 100 last hashes for every partition).